performance/src/main/java/org/hibernate/validator/performance/simple/MultiLevelContainerValidation.java

```diff
@@ -0,0 +1,273 @@
+		Set<ConstraintViolation<MapContainer>> violations = state.validator.validate( state.mapContainer );
+		bh.consume( violations );
+	}
+
+	@Benchmark
+	@BenchmarkMode(Mode.All)
+	@OutputTimeUnit(TimeUnit.MILLISECONDS)
+	@Fork(value = 1)
+	@Threads(50)
+	@Warmup(iterations = 10)
+	@Measurement(iterations = 50)
+	public void testMultiLevelUnConstraintRandomLengthValidation(MultiLevelContainerState state, Blackhole bh) {
+		Set<ConstraintViolation<UnConstraintMapContainer>> violations = state.validator.validate(
+				new UnConstraintMapContainer( prepareTestData( state.random.nextInt( MAX_MAP_ENTRIES ), state.random.nextInt( MAX_LIST_ENTRIES ) ) )
+		);
+		bh.consume( violations );
+	}
+
+	@Benchmark
+	@BenchmarkMode(Mode.All)
+	@OutputTimeUnit(TimeUnit.MILLISECONDS)
+	@Fork(value = 1)
+	@Threads(50)
+	@Warmup(iterations = 10)
+	@Measurement(iterations = 50)
+	public void testMultiLevelRandomLengthValidation(MultiLevelContainerState state, Blackhole bh) {
+		Set<ConstraintViolation<MapContainer>> violations = state.validator.validate(
+				new MapContainer( prepareTestData( state.random.nextInt( MAX_MAP_ENTRIES ), state.random.nextInt( MAX_LIST_ENTRIES ) ) )
+		);
+		bh.consume( violations );
+	}
+
+	@Benchmark
+	@BenchmarkMode(Mode.All)
+	@OutputTimeUnit(TimeUnit.MILLISECONDS)
+	@Fork(value = 1)
+	@Threads(50)
+	@Warmup(iterations = 10)
+	@Measurement(iterations = 50)
+	public void testMultiLevelRandomLengthLessConstraintsValidation(MultiLevelContainerState state, Blackhole bh) {
+		Set<ConstraintViolation<OtherMapContainer>> violations = state.validator.validate(
+				new OtherMapContainer( prepareOtherTestData( state.random.nextInt( MAX_MAP_ENTRIES ), state.random.nextInt( MAX_LIST_ENTRIES ) ) )
+		);
+		bh.consume( violations );
+	}
+
+	/**
+	 * Contains test classes and test data generator.
+	 */
+	static final class MultiLevelContainerDataGenerator {
+
+		private static final Random RANDOM = new Random();
+
+		private MultiLevelContainerDataGenerator() {
+		}
+
+		static Map<Optional<Cinema>, List<EmailAddress>> prepareTestData(int maxEntries, int maxListEntries) {
+			Map<Optional<Cinema>, List<EmailAddress>> map = new LinkedHashMap<>();
+			for ( int i = 0; i < maxEntries; i++ ) {
+				map.put( Optional.of( Cinema.get() ), EmailAddress.getList( maxListEntries ) );
+			}
+			return map;
+		}
+
+		static Map<Cinema, List<EmailAddress>> prepareOtherTestData(int maxEntries, int maxListEntries) {
+			Map<Cinema, List<EmailAddress>> map = new LinkedHashMap<>();
+			for ( int i = 0; i < maxEntries; i++ ) {
+				map.put( Cinema.get(), EmailAddress.getList( maxListEntries ) );
+			}
+			return map;
+		}
+
+		private static String randomString() {
+			char[] chars = new char[RANDOM.nextInt( 10 ) + 1];
+			for ( int i = 0; i < chars.length; i++ ) {
+				chars[i] = (char) RANDOM.nextInt();
+			}
+			return String.valueOf( chars );
+		}
+
+		/**
+		 * Model classes.
+		 */
+
+		static class MapContainer {
```

Comment:
I think these classes should be outside the generator class. It's weird to include them in the generator class.